packages/languages/src/node/language-server-contribution.ts

```diff
@@ -48,27 +72,32 @@ export abstract class BaseLanguageServerContribution implements LanguageServerCo
         return message;
     }
 
-    protected createProcessSocketConnection(
-        outSocket: MaybePromise<net.Socket>, inSocket: MaybePromise<net.Socket>,
-        command: string, args?: string[], options?: cp.SpawnOptions
-    ): Promise<IConnection> {
+    protected async createProcessSocketConnection(outSocket: MaybePromise<net.Socket>, inSocket: MaybePromise<net.Socket>,
+        command: string, args?: string[], options?: cp.SpawnOptions): Promise<IConnection> {
+
         const process = this.spawnProcess(command, args, options);
-        return Promise.all([
-            Promise.resolve(outSocket),
-            Promise.resolve(inSocket)
-        ]).then(result => createProcessSocketConnection(process, result[0], result[1]));
+        const [outSock, inSock] = await Promise.all([outSocket, inSocket]);
+        return createProcessSocketConnection(process.process, outSock, inSock);
     }
 
     protected createProcessStreamConnection(command: string, args?: string[], options?: cp.SpawnOptions): IConnection {
         const process = this.spawnProcess(command, args, options);
-        return createProcessStreamConnection(process);
+        return createStreamConnection(process.output, process.input, () => process.kill());
     }
 
-    protected spawnProcess(command: string, args?: string[], options?: cp.SpawnOptions): cp.ChildProcess {
-        const serverProcess = cp.spawn(command, args, options);
-        serverProcess.once('error', this.onDidFailSpawnProcess.bind(this));
-        serverProcess.stderr.on('data', this.logError.bind(this));
-        return serverProcess;
+    protected spawnProcess(command: string, args?: string[], options?: cp.SpawnOptions): RawProcess {
+        const rawProcess = this.processFactory({ command, args, options });
+        this.disposables.push({
```

Comment:
disposable should be removed if the process exits before stop. `DisposableCollection.push` returns `Disposable` to do it